dmoj/judge.py

```diff
@@ -284,22 +223,238 @@ def listen(self):
+            except Exception:
+                print("Parent process unexpectedly hung up!", file=sys.stderr)
+                traceback.print_exc()
+                self._do_abort()
+                return
+
+            if ipc_msg is None:
+                return
+
+            type, data = ipc_msg
+            if type == IPC_REQUEST_ABORT:
+                self._do_abort()
+            else:
+                raise RuntimeError("unexpected IPC message %s" % str(ipc_msg))
+
+    def _run_grading_wrapper_process(self, conn):
+        self.process_conn.close()
+
+        comm_thread = None
+        try:
+            comm_thread = threading.Thread(target=self._run_judge_request_handler_thread, args=(conn,), daemon=True)
+            comm_thread.start()
+
+            for type, data in self._block_and_grade_cases():
+                conn.send((type, data))
+
+            # Signal to the other end that we're done.
+            conn.send(None)
+        except BrokenPipeError:
+            # There's nothing we can do about this... the general except branch would just fail again. Just re-raise and
+            # hope for the best.
+            raise
+        except:  # noqa: E722, we explicitly want to notify the parent of everything
+            # We can't pickle the whole traceback object, so just send the formatted exception.
+            exc = sys.exc_info()
+            message = ''.join(traceback.format_exception(*exc))
+            conn.send((IPC_UNHANDLED_EXCEPTION, (message,)))
+            conn.send(None)
+        finally:
+            if comm_thread is not None:
+                comm_thread.join(timeout=5)
+
+    def _block_and_grade_cases(self):
+        problem = Problem(self.problem_id, self.time_limit, self.memory_limit, self.meta)
+
+        if 'signature_grader' in problem.config:
+            grader_class = graders.SignatureGrader
+        elif 'interactive' in problem.config:
+            grader_class = graders.BridgedInteractiveGrader
+        elif 'custom_judge' in problem.config:
+            grader_class = graders.CustomGrader
+        else:
+            grader_class = graders.StandardGrader
+
+        try:
+            self.grader = grader_class(self, problem, self.language, utf8bytes(self.source))
+        except CompileError as compilation_error:
+            error = compilation_error.args[0] or b'compiler exited abnormally'
+            yield IPC_COMPILE_ERROR, (error,)
+            return
+        else:
+            binary = self.grader.binary
+            if hasattr(binary, 'warning') and binary.warning is not None:
+                yield IPC_COMPILE_MESSAGE, (binary.warning,)
+
+        yield from self._grade_cases(self.grader, self.grader.cases(), short_circuit=self.short_circuit)
+
+    def _grade_cases(self, grader, cases, short_circuit=False, is_short_circuiting=False, in_batch=False, case_number=0):
+        if not in_batch:
+            yield IPC_GRADING_BEGIN, (self.grader.is_pretested,)
```

Comment:
I am beginning to think it might be better to yield these in `_block_and_grade_cases`. It's annoying that we don't yield `IPC_GRADING_END` when aborted, however.

Reply:
It was like that before, and IMO it was much worse to follow because you had those `BatchBegin`/`BatchEnd`/`TerminateGrading` sentinels.

Reply:
Rewrote all of this.